### ISSUE TITLE
Cross-Record Communication

### ISSUE THREAD

**Issue by bpfarmer** [mir-core/mips#9]
Following #4, I think that it's worth examining whether and how we might enable cross-record communication for stable, mutable records. 

Two examples of where this might be useful are:

1. An on-chain address that could be updated to "recover" from a lost key.

2. An oracle that could provide something like an interest rate or a exchange rate to a stablecoin.

In classic Zexe, the solution to cross-record communication is simple: a consuming record has access to the state of the record that it consumes. However, if we decide to support stable, mutable records, they aren't consumed, so how can they share state?

One possible answer could be that each block has access to the complete state of the stable records in the preceding block. In order to satisfy a mutation predicate in one record that relies on the state of another record, a transaction would need to give a SNARK computing the proof of inclusion of a particular value within the mutable set that matches a particular prefix and corresponds to the block header of the latest block

The problem with this approach is that it requires that a transaction be inserted in the next block, as the proof will be stale otherwise. Another possible solution would be to delegate the computation of the proof of inclusion to miners, and just include the value of a record, such that the block creator could generate the necessary proof based on the preceding block, with some timeout included in the submitted transaction.

**Comment by dlubarov** [comment]
> One possible answer could be that each block has access to the complete state of the stable records in the preceding block.

What if nobody else was interacting with the mutable record I needed though? I might need to "poke" the record myself, but for that I would need the record's state and witness. And if I have that data, then it seems like I could take a more direct approach, and prove the state of the record based on a (perhaps slightly old) root of the commitment set.

> Another possible solution would be to delegate the computation of the proof of inclusion to miners, and just include the value of a record, such that the block creator could generate the necessary proof based on the preceding block, with some timeout included in the submitted transaction.

Hm this raises a good point related to #2. If we had archival nodes, then a user could request the state of a mutable record, like the interest rate in (2), along with its witness. Then the user could prove their transaction valid based on a (slightly old) root of the commitment set, without leaking which mutable records it depended on.

But if we made miners responsible for state proofs, that seems bad for privacy, since a transaction in the mempool would need to declare any mutable records it depended on. And what if, say, my transaction was only valid if the interest rate in (2) is within a certain range? Without archival nodes, I might not have a way to check the current interest rate, so I might not really know whether my transaction is valid, unless I just submit it and see whether a miner picks it up.

**Comment by bpfarmer** [comment]
> What if nobody else was interacting with the mutable record I needed though? I might need to "poke" the record myself, but for that I would need the record's state and witness. And if I have that data, then it seems like I could take a more direct approach, and prove the state of the record based on a (perhaps slightly old) root of the commitment set.

Yeah, I was thinking that transactions that modified mutable records would be "messages" or not records in the usual sense, but you're right that it makes more sense to have them as records in the commitment sets, since the commitment set is a chronologically-ordered Merkle tree, so the worst case is that the proof will need to be appended in the case of an older root. The mutable set would be a prefix tree, so potentially none of the proof of inclusion would be valid for later blocks. 

> But if we made miners responsible for state proofs, that seems bad for privacy, since a transaction in the mempool would need to declare any mutable records it depended on. And what if, say, my transaction was only valid if the interest rate in (2) was within a certain range? Without archival nodes, I might not have a way to check the current interest rate, so I might not really know whether my transaction is valid, unless I just submit it and see whether a miner picks it up.

Yeah I agree. Delegation schemes like this might be worth supporting eventually, as I think that they may make sense for some cases, but you're right on the drawbacks w.r.t. privacy.

**Comment by bpfarmer** [comment]
> What if nobody else was interacting with the mutable record I needed though? I might need to "poke" the record myself, but for that I would need the record's state and witness. And if I have that data, then it seems like I could take a more direct approach, and prove the state of the record based on a (perhaps slightly old) root of the commitment set.

Thinking about this a bit more, it would be nice if we could perhaps use an accumulator to avoid stale witnesses. Here's what I mean in a bit more detail:

Let's say that I want to expose the state of a mutable record - for example, an account where I want to use an associated public key to authenticate something. I'm operating under the assumption that the mutable record state will be stored as a Merkle prefix tree, to enforce uniqueness of bindings between addresses and states. Further, let's suppose that I have a proof-of-inclusion at some block `n` showing that the state of the record corresponding to my account was `v`. As soon as block `n+1` is released, potentially none of that proof-of-inclusion is valid anymore, as any number of leaves may have updated their values.

This isn't ideal, since it means that I'll have to ping a storage node to get an updated proof of inclusion for block `n+k` for all `k`. We could have a separate, chronologically-ordered Merkle tree that stores the value of a record, like the one used for the commitment set, but doesn't save us: I have an old, stale proof-of-inclusion in the prefix tree corresponding to the set of mutable records, so I can't prove directly that the value of the record is `v`. Further, with a chronologically-ordered Merkle tree, I would only have a witness at block `n`, which would be fine, since we could assume that all nodes could store witnesses for the last `k` blocks for reasonably-sized `k`, which would allow them to update my proof of inclusion in the chronologically-ordered tree. But, updating the value of a record would require updating the chronologically-ordered tree, which would void all past witnesses. A solution to this could be a proof-of-exclusion in the nullifier set, but this would likely be a Merkle prefix tree, and therefore would require updating on each new block.

This is a long-winded way of saying that it might be nice to have an accumulator that is *dynamic* in the sense given [here](https://eprint.iacr.org/2018/1188) such that I could efficiently prove that a particular value is "alive" and bound to a particular record in the latest block without having to download an entirely new proof of inclusion, just an old proof of inclusion in the mutable set's prefix tree. This might not actually be worth worrying about, but an interesting problem to consider.

**Comment by dlubarov** [comment]
> updating the value of a record would require updating the chronologically-ordered tree, which would void all past witnesses

These are good points, and I think if we don't use nullifiers, they also apply to immutable records. I.e. marking an immutable record as consumed/deleted would also void other witnesses.

> A solution to this could be a proof-of-exclusion in the nullifier set, but this would likely be a Merkle prefix tree, and therefore would require updating on each new block.

Yeah nullifiers could work, but I think storage-wise, it would be better to store a hash of each live record. So if a record underwent `n` mutations, at the end we'd still have one hash, versus `n` nullifiers.

Another idea is to have an (append only) chronologically-ordered Merkle tree containing record snapshots, each with a version number, and then a second (updatable) tree containing the current version number of each mutable. To prove the current state of a mutable record, you'd prove the inclusion of its snapshot in the first tree, then prove that that snapshot's version number was current using the second tree. The network would have to store the full second tree, but the version numbers could use a compact variable-length integer representation, so they could be much smaller than hashes, especially for records with relatively few updates.

Alternatively, what do you think of having the network store one accumulator for each user? When a user mutated one of their records, they'd update their other witnesses, but no external users would need to do so. Of course we couldn't prevent someone from creating multiple accumulators, but we could charge a nontrivial fee to discourage it. `32 bytes/user * 7 billion users` is 224 gb, which seems pretty manageable. Unless we had a lot of public records with no owner; this wouldn't help with that.

> it might be nice to have an accumulator that is _dynamic_ in the sense given [here](https://eprint.iacr.org/2018/1188)

Haven't had a chance to read it yet but I'll take a look soon!
